### DISCUSSION TITLE
data uploading pipeline

### DISCUSSION THREAD

**Discussion by JinIgarashi** [UNDP-Data/geohub#545] [Ideas]
@iferencik I made an picture how we can establish data uploading pipeline in Azure.

![data-upload drawio](https://user-images.githubusercontent.com/2639701/179242516-08e99354-db74-4923-9616-c91418f5ff61.svg)

- References
   - How to add queue in Azure Service Bus from Azure Function
https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-service-bus-output?tabs=in-process%2Cextensionv5&pivots=programming-language-javascript
   - How to trigger Azure Function from Azure Service Bus
https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-service-bus-trigger?tabs=in-process%2Cextensionv5&pivots=programming-language-javascript

**Comment by JinIgarashi** [comment]
According to the discussion with @iferencik, I updated the image. First Azure Function to upload data can be replaced svelte endpoint. Also, Azure Blob can trigger Azure Function

![data-upload](https://user-images.githubusercontent.com/2639701/181756430-3a6bc15e-3f8a-40c9-b528-e46ad9ce56f9.png)

References
- Azure Blob storage trigger for Azure Functions
   https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-storage-blob-trigger?tabs=in-process%2Cextensionv5&pivots=programming-language-javascript

**Comment by JinIgarashi** [comment]
## Target of uploaded data format
Shorlist some drivers, accept  ALL GDAL supported
### Raster

- TIFF
- netCDF
- Arc/Info ASCII Grid
- Erdas Imagine



### Vector
- ESRI Shapefile (ZIP)
- GeoJSON
- PMTILES
- MBTILES



### GeoDatabases
- ESRI File Geodatabase (OpenFileGDB)
- ESRI File Geodatabase (FileGDB)
- ESRI Personal GeoDatabase
- GeoPackage

### Uploading approaches

- directly upload COG or pmtiles to Azure Blob (no need additional processing
- upload original GIS data to Azure Blob, then docker container based service (AKS) will convert data to either COG or pmtiles on Azure Blob

### Blob container

- a user container (named by user email address) to manage all files uploaded by user
- every user container should have the following metadata tags
  - published=true
  - label=User Name (e.g., Jin Igarashi's container)
  - description=description for user container (e.g., Jin Igarashi's container to manage uploaded data)
  - icon=no need to set
  - tags=no need to set

folder structure can be....

```
|- `jin.igarashi@undp.org`
  |- tmp: folder which stores original data for data processing
  |  |- test.zip: shapefile
  |- test.pmtiles - after converting test.zip
  |- aaa.pmtiles - directly uploaded by frontend
|- `ioan.ferencik@undp.org`
|- ...
```

- if data is not COG or pmtiles, create `original` folder under user folder
  - call services running on AKS to process data to convert to cloud optimised format
  - AKS service will save converted file under user folder
  - call Azure Function service to register metadata to database.

- if data is already COG or pmtiles
  - check metadata is appropriate. COG's metadata are all available, is there tilestats, vectorinfo property in pmtiles
  - saved file to user folder
  - call Azure Function service to register metadata to database.

**Reply by JinIgarashi** [reply]
This is copy from draft PR#1250

![geohub-database-erd](https://user-images.githubusercontent.com/2639701/215550604-650f25df-3d12-47ad-8dc5-23332ec4a227.png)


The following procedure can be done associated with this table structure

#### data access management

Regarding to access level for datasets, there are two types.
- the range of data accessibility by users
  - I added `access_level` column in `geohub.dataset` table for this feature.
  - the values can be the same with `style` table
    - 1: login user only can access from GeoHub
    - 2: UNDP user and login user only can access from GeoHub
    - 3: anonymous users can access from GeoHub
- the access permission for data management by users
  - I added `geohub.dataset_access` table for this feature
  - we can add login user as admin permission as default
  - `permission` field can be, 0: no any right to change this dataset , 1: read only, 2: read/write, 3: owner
  - `user` column can have user's email address which can access to change this datasets
  - `owner` user can invite multiple users to manage this dataset
  -  read only users can see metadata information from data management portal
  -  read/write users can edit metadata from data management portal
  - owner of datasets can do the same things with read/write users and also can delete datasets

#### data uploading and processing management

I created `geohub.dataset_processing` table for managing status of data processing

1. user upload GIS data to Azure Blob contaner
2. registered dataset in the following three tables

- `storage` table
  - If user's storage is not yet registered, insert it to the table
  - create user's container in Azure Storage account (container name is user email address)
- `dataset` table
  - registered information
    - id: id can be generated proposed dataset URL
    - storage_id
    - url (proposed cloud optimised dataset URL.)
    - name: given by frontend
    - description: givern by frontend
    - is_raster: decided by file format uploaded
    - license: givern by frontend
    - bounds: NULL value for this stage
    - createdat: date of uploaded
    - access_level: given by frontend
- geohub.dataset_access
  - registered information
    - dataset_id
    - permission: 3 (owner)
    - user: login user's email address
    - createdat: datetime of uploaded
 - geohub.dataset_processing
   - registered information
     - dataset_id
     - processedat: datatime of data uploading
     - data_url: original data's blob URL to be processed
     - status: 1 (data uploaded)
     - user: login user email address

3. pipeline (maybe AKS) will be triggered once a record is inserted to data_processing table
4. pipeline update `geohub.data_processing` table
- insert a record as following information
  - dataset_id
  - processedat: datetime started processing data
  - data_url: NULL
  - status: 2 (started processing)
  - user: app name of pipeline
5. pipeline update datasets table with new information after data processing will be done
- Update following information for dataset table
  - `bounds`: update bounds by using the metadata after data is processed
  - `updatedat`: updated datetime
  - Once bounds is set, `/datasets` api can search data
- insert a record for `dataset_processing` table
  - dataset_id
  - processedat: datetime ended processing
  - data_url: null
  - status: 3 (data processing ended)
  - user: app name of data processing

**Reply by JinIgarashi** [reply]
additional format should be supported?

- Esri file geodabase (vector and raster)
- geopackage (vector and raster)

**Reply by iferencik** [reply]
## Context
The data management functionality in GeoHub is going to support users with (1) uploading (2) publishing and (3) managing their geo-spatial data. The main goal here is to break the geo-spatial data silos within UNDP. A valid assumption is that significant quantities of geo-spatial data sits isolated across the organization and users could help solve this issue by publishing this data through GeoHub.

While some users might have geo-spatial background and thus know more about the data it is only safe to assume  most of users have little or no  knowledge  at all about  GIS data formats. Another assumption is the users will in general posses a proprietary GIS format from ESRI or MapInfo that will hold multiple GIS layers.

## User journey
Only authenticated users will be able to upload GIS data to GeoHub.
The data management functionality will  be powered by Azure Blob storage which constitutes a adequate storage environment. The user journey start with uploading  geospatial data. It then continues with the datasets table, a page that shows a list of data files/data bases  and datasets.The table shows every uploaded geo-spatial data file or database including some of its properties like  file name, file size in GB, and status (uploaded, ingesting, ingested). As soon as the status of  an uploaded data file/database is ingested the user can  visualize the available dataset that exist inside the data file.
By default no available dataset is published. The user can choose to publish any available dataset and  for this is taken to a new page. This is necessary because every dataset , in order to be publishable needs to contain  a minimum requited set of metadata like description, tags, etc. Once the metadata is filled in the user is taken back to the datasets table and the status of the dataset is changed to published. Next the user can un-publish or completely delete a given dataset.




![image](https://user-images.githubusercontent.com/92793531/218161964-25bb1bb4-e777-42dd-98d6-d575e1475bca.png)

![image](https://user-images.githubusercontent.com/92793531/218166566-ef2c5755-bbad-41b6-b5bd-bbe579f86266.png)

```mermaid
flowchart LR
A(Authenticated user) -->B(Datasets table) --> C(Upload new)--file-->B(Datasets table)--Dataset-->D(Publish)-->B(Datasets table)
B--Dataset-->E(un-publish)
B--Dataset-->F(Remove)
```
## Data journey
For the users to be able to manage their own GIS datasets the Geohub must implement functionality on the front-end as well as on the back-end side. The starting point of this process  is the user uploading a data file.  By convention, user's data will be hosted in a specific Azure blob storage container inside a folder carrying the email address of the user *firstname.lastname@undp.org*. However, this  could affect the final path of the server datasets and the ```@``` character could potentially break software/services/clients. On solution could be to hash this string using  md5 algorithm and use that value for the folder name.

Inside this folder three sub-folders named **raw**, **working**, and **datasets** will be created and will serve the ingest process. The **raw** folder is the target for the upload process and it will hold the original versions of user uploaded data files. These data files will stay here until the owner of an administrator decides to remove it. The second folder **working** is meant to hold a copy of the uploaded data files temporarily, until they are ingested. The option to keep the copied data files permanently is also a possibility here, however this could have an impact on the size of the data base. The last folder **datasets** is the folder where all the user's datasets are going to be stored. The GeoHub layers extracted from any given uploaded data file are going to be stored inside this folder in a sub-folder with the same name as the input data file. For example if the user has uploaded a data file to:

**firstname.lastname@undp.org/raw/afile.tif** the resulting layers/GeoHub dataset would be located in

**firstname.lastname@undp.org/datasets/afile.tif/band1.tif**


The path of a geospatial archive/data file is as follows




```mermaid
flowchart LR
    subgraph FRONTEND
        direction LR
        A[Upload]--file-->B[userdata/userid/raw]
    end
    subgraph BACKEND
        direction TB
        M[userdata/userid/raw]-->D{Sanity Check}
        D{Sanity Check} -- Fail --> E[userdata/userid/corrupted] ----> E1{Manual check} 
        E1{Manual check}  --Fail--> E2[DELETE] --> E3[Email]
         E1{Manual check}  --Pass-->M[userdata/userid/raw]
         F[userdata/userid/working] --create folder --> C[userdata/userid/datasets/datafile]
        D{Sanity Check} -- Pass --> F[userdata/userid/working] --Check data type-->G{Data type}
        G{Data type} --raster--> R[raster ingest] --COG--> R1(userdata/userid/datasets/datafile/layername.tif)
         G{Data type} --vector--> V[vector ingest] --PMTILES--> V1(userdata/userid/datasets/datafile/layername.pmtiles)
          G{Data type} --CSV--> Z[CSV ingest] --COG/PMTILES--> Z1(userdata/userid/datasets/datafile/layername.tif?pmtiles)
       
    end

    
    FRONTEND --> |BLOB TRIGGER| BACKEND

   
```

**Comment by iferencik** [comment]
**My workspace is a section of GeoHub that allows users to manage their data.**

| Data file name |              Status | Size | Layers|
|:-----------|:-----------|:-------:|:-----------:|
| a long_Name-File.gdb      | uploaded    | 0.4GB|**&#43258;**|
| Paragraph   | Text        |
</center>
- [ ] 1. shows the list of existing data files-> layers

1. 

   every data file becomes a line with buttons to delete the data file and columns to show the file size in MB/GB, uploaded
   


2.uploading $ingest
3.editing metadata for ingested layers
4.publishing to geoHub
5.un-publishing
6.deleting datasets/layers
7.deleting data files (including all associated datasets/layers)

**Comment by jessefriend** [comment]
The individual user folder structure in the storage account 
|-userdata - complete storage account
   |- `user-hash`
      |- `raw`: folder which stores upload data
        - example_multiband.tif -file
      |- `working` - original file copied to this folder to do ingesting on
        - example_multiband.tif - same file
      |- `datasets`
        - `example_multiband.tif` - subdirectory
            -example_multiband_b01.tif - file
            -example_multiband_b02.tif - file

**Reply by JinIgarashi** [reply]
- raw
  - aaa_yyyymmddhhmmss.tif - original data
  - aaa_yyyymmddhhmmss.tif.error - in case data processing failed. error file, uploaded if error occurs before dataset is created.
  - aaa_yyyymmddhhmmss.tif.log - a log file representing all the logs generated by the pipeline just as it was ingested
- datasets
  - aaa_yyyymmddhhmmss.tif - subfolder
    - aaa_yyyymmddhhmmss.tif - processed data
    - aaa_yyyymmddhhmmss.tif.ingesting - processing end, but unpublished

**Comment by JinIgarashi** [comment]
This is the proposed JSON data which the pipeline is going to send to sveltekit websocket server.

```json
{
  "user": "a85516c81c0b78d3e89d3f00099b8b15",
  "url": "https://undpgeohub.blob.core.windows.net/userdata/a85516c81c0b78d3e89d3f00099b8b15/raw/wasac-rwasom-2-data-revised_20230912143433.gpkg",
  "progress": 10,
  "stage": "downloading"
}
```

- `user` property should be the MD5 hash generated from user email address for the dataset
- `url` property should be a raw file blob URL of storage account.
- `progress` property should an integer number which is ranged from 0 to 100 to show progress.
- `stage`:
  - provisioning infrastructure
  - downloading
  - ingesting
  - uploading

**Reply by JinIgarashi** [reply]
Message format should be as following.

```json
{
  "type": "message",
  "from": "server",
  "group":  "datapipeline",
  "dataType": "json",
  "data": {
    "user": "a85516c81c0b78d3e89d3f00099b8b15",
    "url": "https://undpgeohub.blob.core.windows.net/userdata/a85516c81c0b78d3e89d3f00099b8b15/raw/wasac-rwasom-2-data-revised_20230912143433.gpkg",
    "progress": 10,
    "stage": "downloading"
  }
}
```

**Reply by JinIgarashi** [reply]
For cancelation
```
{
  "type": "message",
  "from": "server",
  "group":  "datapipeline",
  "dataType": "json",
  "data": {
    "user": "a85516c81c0b78d3e89d3f00099b8b15",
    "url": "https://undpgeohub.blob.core.windows.net/userdata/a85516c81c0b78d3e89d3f00099b8b15/raw/wasac-rwasom-2-data-revised_20230912143433.gpkg",
    "cancel": true
  }
}
```

**Comment by JinIgarashi** [comment]
The message will be sent to sevice bus queue when users upload the dataset, it follows the following message format.

```
${blobUrl};${token};join_vectortiles={true/false}
```

each infomation is split by `;`.

- blobUrl: uploaded dataset's URL
- token: access token from Auth.js (Azure AD or GitHub)
- join_vectortiles: if true, the pipeline will generate only a single PMTiles with all vector layers.

**Comment by iferencik** [comment]
**Feature set progress state as blob metadata**

The geohub-data-pipeline  communicates with the GeoHub UI through AsurePubSub  Websocket service. The pipeline sends messages with the status and progress on every data file. While this keeps the UI nad pipeline in sync if the user refreshed the page while a dataset is being processed all the progress and status is lost. For this purpose everytime a websocket message is sent the stage and progress is set on the  raw datafile  as metadata.

**Comment by Thuhaa** [comment]
Feature: Upload multiple datasets to Geohub
1. This is a currently already implemented feature, where users can upload multiple datasets at the same time to GeoHub
2. To do this, the user simply needs to select multiple files at the same time on GeoHubs upload data page. https://geohub.data.undp.org/data/upload or https://undpgeohub.com/data/upload.
3. The selected files are going to be zipped and uploaded to the blob storage as zipfiles
4. Moreover, there is now a check for shapefiles that check if the selected shapefile is accompanied by the required files (`dbf`, `shx`, `shp`, `prj`). This check however does not guarantee that the shapefile is valid and will pass GDAL/OGR checks so that the data can be ingested by GeoHub. Users can check the logs on the data page for any data errors not caught.

**Comment by JinIgarashi** [comment]
let's close this discussion since it has been implemented
